GAE/src/org/waterforpeople/mapping/serialization/SurveyInstanceHandler.java

```diff
@@ -86,27 +91,58 @@ public static SurveyInstance fromJSON(String data) {
             } else {
                 si.getQuestionAnswersStore().add(qas);
             }
-            
+
         }
-        
+
         return si;
     }
-    
+
+    /*
+     * Retrieve the formId for form data coming in via data.json files. This extra check for the
+     * formId has been created in order to handle a bug where the formId was returned as an
+     * arbitrary string. We ensure that the formId is strictly a long otherwise we try to retrieve
+     * it through the questionId values that come with each individual question response. See
+     * https://github.com/akvo/akvo-flow-mobile/issues/614
+     */
+    private static Long retrieveFormId(FormInstance formInstance) {
+        if (StringUtils.isNotBlank(formInstance.getFormId())
+                && StringUtils.isNumeric(formInstance.getFormId())) {
+            return Long.parseLong(formInstance.getFormId());
+        }
+
+        for (Response response : formInstance.getResponses()) {
+            Long formId = retriveFormIdByQuestionId(response.getQuestionId());
+            if (formId != null) {
+                return formId;
+            }
+        }
+
+        return null;
+    }
+
+    /*
+     * Retrieve the formId from the corresponding Question entity in the datastore
+     */
+    private static Long retriveFormIdByQuestionId(String questionId) {
+        Question question = new QuestionDao().getByKey(Long.parseLong(questionId));
```

Comment:
is it possible that questionId is null or empty? or can it fail because this question has been deleted?